emds: Revise rram related timing in emds (#19275)
nrfconnect/sdk-nrf: doc/nrf/libraries/others/emds.rst

However, this is chip-dependent, so it is important to measure the time.
Find timing values under the "Electrical specification" section for the non-volatile memory controller in the Product Specification for the relevant SoC or the SiP you are using.
For example, for the nRF52840 SiP, see the `nRF52840 Product Specification`_ page.

The following Kconfig options can be configured:

* :kconfig:option:`CONFIG_EMDS_FLASH_TIME_WRITE_ONE_WORD_US`
* :kconfig:option:`CONFIG_EMDS_FLASH_TIME_ENTRY_OVERHEAD_US`
* :kconfig:option:`CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US`

When configuring these values, consider the time for erase when doing garbage collection in NVS using flash.
If partial erase is not enabled or supported, the time of a complete sector erase has to be included in the :kconfig:option:`CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US`.
When partial erase is enabled and supported by the hardware, include the time it takes for the scheduler to trigger, which is depending on the time defined in :kconfig:option:`CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE_MS`.
When changing the :kconfig:option:`CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US` option, it is important that the worst time is considered.

> When configuring these values using RRAM there is not a need to consider garbage collection in the same way as for flash, as it can be written to regardless of value. But the ERASEALL functionality

Reviewer: ```suggestion
When configuring these values using RRAM, you do not need to consider garbage collection in the same way as for flash, as it can be written to regardless of the value.
However, avoid the ERASEALL functionality,
```

> should be avoided as that can increase the time before the EMDS store functions are called.

Reviewer: ```suggestion
because that can increase the time before the EMDS store functions are called.
```

> Overview
> ********

Reviewer: ```suggestion
```

> The emergency data storage (EMDS) library provides persistent storage functionality designed to prevent the wear and tear of the flash/RRAM.

Reviewer 2: ```suggestion
The emergency data storage (EMDS) library provides persistent storage functionality designed to prevent the wear and tear of the flash memory or RRAM.
```

Ideally change to this everywhere it appears, or something else perhaps, but I think "flash/RRAM", especially considering how often it appears in this file, breaks the flow of the text quite a bit, so I hope we can find something that works a bit better.

I have added several other comments for specific places where I have other suggested rewordings for those specific sentences.

Reviewer 2: @peknis What are your thoughts on this? Any ideas for how to improve the flow a bit, or is it fine as it is?

Reviewer: Otherwise your suggestion looks good, but you could have this intro before the Overview section. If there is nothing else noteworthy to tell about the library, you can leave out that section altogether. See https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/templates/library_template_README.html for further tips and hits on writing a library doc.

Its intended use is for storing of data undergoing frequent changes during runtime.

> Implementation
> ==============
> The :kconfig:option:`CONFIG_EMDS` Kconfig option enables the emergency data storage.

Reviewer: ```suggestion
Implementation
**************

The :kconfig:option:`CONFIG_EMDS` Kconfig option enables the emergency data storage.
```

> The application must initialize the pre-allocated flash/RRAM area by using the :c:func:`emds_init` function.

Reviewer 2: ```suggestion
The application must initialize the pre-allocated storage area by using the :c:func:`emds_init` function.
```

The :kconfig:option:`CONFIG_EMDS_SECTOR_COUNT` option defines how many sectors should be used to store data.

The allocated storage space must be larger than the combined data stored by the application.

> Allocating a larger flash/RRAM area will demand more resources, but also increase the life expectancy of the flash/RRAM.

Reviewer 2: ```suggestion
Allocating a larger storage area will demand more resources, but also increase the life expectancy of the flash/RRAM.
```

> Allocating a larger storage area will demand more resources, but also increase the life expectancy of the flash/RRAM.

Reviewer 2: My idea was to replace the remaining "flash/RRAM" with "flash or RRAM", I'm not really a fan of the slashes. Up to you, won't block on this.

The chosen size should reflect the amount of data stored, the available flash/RRAM resources, and how the application calls the :c:func:`emds_store` function.

> In general, it should not be necessary to allocate a large flash/RRAM area, since only a limited set of data should be stored to ensure swift completion of writing the flash/RRAM on shutdown.

Reviewer 2: ```suggestion
In general, it should not be necessary to allocate a large storage area, since only a limited set of data should be stored to ensure swift completion of writing the data on shutdown.
```

The memory location that is going to be stored must be added on initialization.
All memory areas must be provided through entries containing an ID, data pointer, and data length, using the :c:func:`emds_entry_add` function and the :c:macro:`EMDS_STATIC_ENTRY_DEFINE` macro.
Entries to be stored when the emergency data storage is triggered need their own unique IDs that are not changed after a reboot.

When all entries are added, the :c:func:`emds_load` function restores the entries into the memory areas from the flash/RRAM.

> After restoring the previous data, the application must run the :c:func:`emds_prepare` function to prepare the flash/RRAM area for receiving new entries.

Reviewer 2: ```suggestion
After restoring the previous data, the application must run the :c:func:`emds_prepare` function to prepare the storage area for receiving new entries.
```

> If the remaining empty flash/RRAM area is smaller than the required data size, the flash/RRAM area will be automatically erased to increase the available flash/RRAM area.

Reviewer 2: ```suggestion
If the remaining empty storage area is smaller than the required data size, the storage area will be automatically erased to increase the available storage area.
```

The storage is done in deterministic time, so it is possible to know how long it takes to store all registered entries.
However, this is chip-dependent, so it is important to measure the time.
Find timing values under the "Electrical specification" section for the non-volatile memory controller in the Product Specification for the relevant SoC or the SiP you are using.
For example, for the nRF52840 SiP, see the `nRF52840 Product Specification`_ page.

The following Kconfig options can be configured:

* :kconfig:option:`CONFIG_EMDS_FLASH_TIME_WRITE_ONE_WORD_US`
* :kconfig:option:`CONFIG_EMDS_FLASH_TIME_ENTRY_OVERHEAD_US`
* :kconfig:option:`CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US`

> When configuring these values, consider the time for erase when doing garbage collection in NVS using flash.

Reviewer 2: NVS is doing garbage collection both with flash and RRAM isn't it? Should this line stay as it was?

Author: I would not add ZMS, as I do not think it is recommended to use ZMS together with flash. That is way I said using NVS together with flash.

If partial erase is not enabled or supported, the time of a complete sector erase has to be included in the :kconfig:option:`CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US`.
When partial erase is enabled and supported by the hardware, include the time it takes for the scheduler to trigger, which is depending on the time defined in :kconfig:option:`CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE_MS`.
When changing the :kconfig:option:`CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US` option, it is important that the worst time is considered.

> When configuring these values using RRAM there is not a need to consider garbage collection in the same way as for flash, as it can be written to regardless of value. But the ERASEALL functionality

Reviewer 3: > But the ERASEALL functionality should be avoided

How can you 'avoid' eraseall functionality? EMDS storage page erase anyway needed upon the next reboot right?

Author: RRAMC provides functionality to protect the device against the erase all function.
See https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/rramc.html#ariaid-title7

> When configuring these values using RRAM there is not a need to consider garbage collection in the same way as for flash, as it can be written to regardless of value. But the ERASEALL functionality

Reviewer 2: Is this true both for NVS and ZMS?

Author: I believe this is true regardless of NVS or ZMS. This is functionality for the RRAM controller so this will be true regardless of what you use on top. ERASEALL can for RRAM be blocked. When using RRAM you do not need to think about garbage collection the same way as for flash memory. How ZMS and NVS is handling this on top of this is different story.

Once the data storage has completed, a callback is called if provided in :c:func:`emds_init`.
This callback notifies the application that the data storage has completed, and can be used to reboot the CPU or execute another function that is needed.

> After completion of :c:func:`emds_store`, the :c:func:`emds_is_ready` function call will return error, since it can no longer guarantee that the data will fit into the persistent memory area.

Reviewer: ```suggestion
After completion of :c:func:`emds_store`, the :c:func:`emds_is_ready` function call will return an error, because it can no longer guarantee that the data will fit into the persistent memory area.
```

Once the data storage has completed, a callback is called if provided in :c:func:`emds_init`.
This callback notifies the application that the data storage has completed, and can be used to reboot the CPU or execute another function that is needed.

> After completion of :c:func:`emds_store`, the :c:func:`emds_is_ready` function call will return error, since it can no longer guarantee that the data will fit into the flash/RRAM area.

Reviewer 2: ```suggestion
After completion of :c:func:`emds_store`, the :c:func:`emds_is_ready` function call will return error, since it can no longer guarantee that the data will fit into the storage area.
```

Reviewer 2: Looks like you forgot this one

Requirements
************

> To prevent frequent writes to flash/RRAM memory, the EMDS library can write data to flash/RRAM only when the device is shutting down.

Reviewer 2: ```suggestion
To prevent frequent writes to persistent memory, the EMDS library can write data only when the device is shutting down.
```

   t_\text{store} = t_\text{base} + \sum_{i = 1}^n \left(t_\text{entry} + t_\text{word}\left(\left\lceil\frac{s_\text{ate}}{s_\text{block}}\right\rceil + \left\lceil\frac{s_i}{s_\text{block}}\right\rceil \right)\right)

where :math:`t_\text{base}` is the value specified by :kconfig:option:`CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US`, :math:`t_\text{entry}` is the value specified by :kconfig:option:`CONFIG_EMDS_FLASH_TIME_ENTRY_OVERHEAD_US` and :math:`t_\text{word}` is the value specified by :kconfig:option:`CONFIG_EMDS_FLASH_TIME_WRITE_ONE_WORD_US`.

> :math:`s_i` is the size of the :math:`i`\ th entry in bytes and :math:`s_\text{block}` is the number of bytes in one word of flash, while for RRAM the write-buffer size i number of 128-bit words (16 bytes).

Reviewer: ```suggestion
:math:`s_i` is the size of the :math:`i`\ th entry in bytes and :math:`s_\text{block}` is the number of bytes in one word of flash, while for RRAM the write-buffer size ``i`` number of 128-bit words (16 bytes).
```

> :math:`s_i` is the size of the :math:`i`\ th entry in bytes and :math:`s_\text{block}` is the number of bytes in one word of flash, while for RRAM the write-buffer size ``i`` number of 128-bit words (16 bytes).

Reviewer 2: This sentence doesn't make sense when I try to read it? Is $s_i =$ the number of 128-bit words used by the entry? Or is $s_\text{block} =$ 16? It would be good if you used the terms from the formula (`` :math:`s_i\` `` and `` :math:`s_\text{block}` ``) and explain what values they should have in the RRAM case (what is "write-buffer size" in this case for instance? I can't tell where to put 128 and/or 16 into the formula)

   t_\text{store} = t_\text{base} + \sum_{i = 1}^n \left(t_\text{entry} + t_\text{word}\left(\left\lceil\frac{s_\text{ate}}{s_\text{block}}\right\rceil + \left\lceil\frac{s_i}{s_\text{block}}\right\rceil \right)\right)

where :math:`t_\text{base}` is the value specified by :kconfig:option:`CONFIG_EMDS_FLASH_TIME_BASE_OVERHEAD_US`, :math:`t_\text{entry}` is the value specified by :kconfig:option:`CONFIG_EMDS_FLASH_TIME_ENTRY_OVERHEAD_US` and :math:`t_\text{word}` is the value specified by :kconfig:option:`CONFIG_EMDS_FLASH_TIME_WRITE_ONE_WORD_US`.
:math:`s_i` is the size of the :math:`i`\ th entry in bytes and :math:`s_\text{block}` is the number of bytes in one word of flash/RRAM.
These can be found by looking at datasheets, driver documentation, and the configuration of the application.

> :math:`s_\text{ate}` is the size of the allocation table entry used by the EMDS flash/RRAM module, which is 8 B.

Reviewer 3: ```suggestion
:math:`s_\text{ate}` is the size of the allocation table entry used by the EMDS flash/RRAM module, which is 8 B if NVS is used, else it is 16 B for ZMS.
```

Author: The ate size is 8 regardless of ZMS or NVS. It is defined in emds_flash.c.

   This is the worst case overhead when a store is triggered in the middle of an erase on nRF52840 with :kconfig:option:`CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE` enabled in the sample by default, and should be adjusted when using other configurations.
*  Set :math:`t_\text{entry}` = 300 µs and :math:`t_\text{word}` = 41 µs. *Note: These values are valid only for this specific chip and configuration, and should be computed for the specific configuration whenever using EMDS.*
*  The sample uses two entries, one for the RPL with 255 entries (:math:`s_i` = 2040 B) and one for the lightness state (:math:`s_i` = 3 B).

> *  The flash/RRAM write block size :math:`s_\text{block}` in this case is 4 B, and the ATE size :math:`s_\text{ate}` is 8 B.

Reviewer 3: ```suggestion
*  The flash/RRAM write block size :math:`s_\text{block}` in this case is 4 B, and for NVS backend the ATE size :math:`s_\text{ate}` is 8 B.
```
